File: alura/php-design-patterns/Aprovado.php
<?php
  class Aprovado implements EstadoDeUmOrcamento {
    private $descontoAplicado = false;

    public function aplicaDescontoExtra(Orcamento $orcamento) {
      if(!$this->descontoAplicado) {
        $orcamento->setValor($orcamento->getValor() - $orcamento->getValor() * 0.05);
        $this->descontoAplicado = true;
      } else {
        throw new Exception("Desconto já aplicado!");
      }
    }

    public function aprova(Orcamento $orcamento) {
      throw new Exception("Este orçamento já se encontra aprovado");
    }

    public function reprova(Orcamento $orcamento) {
      throw new Exception("Um orcamento aprovado não pode ser reprovado");
    }

    public function finaliza(Orcamento $orcamento)  {
      $orcamento->setEstado(new Finalizado());
    }
  }
?>

File: alura/c-sharp/design-patterns-1/Aprovado.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CursoDesignPatterns
{
  public class Aprovado : EstadoDeUmOrcamento
  {
    public void AplicaDescontoExtra(Orcamento orcamento)
    {
      orcamento.Valor -= orcamento.Valor * 0.02;
    }

    public void Aprova(Orcamento orcamento) {
      throw new Exception("Orcamento já está em estado de aprovacao");
    }

    public void Reprova(Orcamento orcamento) {
      throw new Exception("Orcamento esta em aprovado, não pode ser reprovado agora");
    }

    public void Finaliza(Orcamento orcamento) {
      orcamento.EstadoAtual = new Finalizado();
    }
  }
}
